packages/backend-test-utils/src/next/wiring/ServiceFactoryTester.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ServiceFactory, ServiceRef } from '@backstage/backend-plugin-api';
+import { defaultServiceFactories } from './TestBackend';
+// Direct internal import to avoid duplication.
+// This is a relative import in order to make sure that the implementation is duplicated
+// rather than leading to an import from @backstage/backend-app-api.
+// eslint-disable-next-line @backstage/no-relative-monorepo-imports
+import { ServiceRegistry } from '../../../../backend-app-api/src/wiring/ServiceRegistry';
+
+/**
+ * Options for {@link ServiceFactoryTester}.
+ * @public
+ */
+export interface ServiceFactoryTesterOptions {
+  /**
+   * Additional service factories to make available as dependencies.
+   *
+   * @remarks
+   *
+   * If a service factory is provided for a service that already has a default
+   * implementation, the provided factory will override the default.
+   */
+  dependencies?: Array<ServiceFactory | (() => ServiceFactory)>;
+}
+
+/**
+ * A utility to help test service factories in isolation.
+ *
+ * @public
+ */
+export class ServiceFactoryTester<TService, TScope extends 'root' | 'plugin'> {
+  readonly #subject: ServiceRef<TService, TScope>;
+  readonly #registry: ServiceRegistry;
+
+  /**
+   * Creates a new {@link ServiceFactoryTester} used to test the provided subject.
+   *
+   * @param subject - The service factory to test.
+   * @param options - Additional options
+   * @returns A new tester instance for the provided subject.
+   */
+  static from<TService, TScope extends 'root' | 'plugin'>(
+    subject:
+      | ServiceFactory<TService, TScope>
+      | (() => ServiceFactory<TService, TScope>),
+    options?: ServiceFactoryTesterOptions,
+  ) {
+    return new ServiceFactoryTester(
+      typeof subject === 'function' ? subject() : subject,
+      options?.dependencies,
+    );
+  }
+
+  private constructor(
+    subject: ServiceFactory<TService, TScope>,
+    dependencies?: Array<ServiceFactory | (() => ServiceFactory)>,
+  ) {
+    this.#subject = subject.service;
+
+    this.#registry = new ServiceRegistry([
+      ...defaultServiceFactories,
+      ...(dependencies?.map(f => (typeof f === 'function' ? f() : f)) ?? []),
```

Comment:
Neater to do this in `from`, just like for the subject? Just to have them side by side. That could also be a place to do the `??` replacement so that the constructor can be just two plain assignments